geoips/interfaces/module_based/readers.py

```diff
@@ -19,5 +24,233 @@ class ReadersInterface(BaseModuleInterface):
+            * List of desired channels (skip unneeded variables as needed).
+            * Include all channels if None.
+        area_def : pyresample.AreaDefinition, default=None
+            * Specify region to read
+            * Read all data if None.
+        self_register : str or bool, default=False
+            * register all data to the specified dataset id (as specified in the
+              return dictionary keys).
+            * Read multiple resolutions of data if False.
+
+        Returns
+        -------
+        dict of xarray.Datasets
+            * dictionary of xarray.Dataset objects with required Variables and
+              Attributes.
+            * Dictionary keys can be any descriptive dataset ids.
+
+        See Also
+        --------
+        :ref:`xarray_standards`
+            Additional information regarding required attributes and variables
+            for GeoIPS-formatted xarray Datasets.
+        """
+        all_metadata = self.concatenate_metadata(
+            [call_single_file_func([x], metadata_only=True)["METADATA"] for x in fnames]
+        )
+        if metadata_only:
+            return all_metadata
+
+        dict_xarrays = self.call_files_and_get_top_level_metadata(
+            fnames,
+            all_metadata,
+            call_single_file_func,
+            metadata_only,
+            chans,
+            area_def,
+            self_register,
+        )
+        return dict_xarrays
+
+    def concatenate_metadata(self, all_metadata):
+        """Merge together metadata sourced from a list of files into one dictionary.
+
+        Where the structure of the merged metadata is a nested dictionary of metadata.
+
+        Ie. (xarray_obj has no data and is merely just a container for metadata):
+        {"METADATA": xobj.source_file_attributes: {fname: xobj, ..., "fnamex": xobj}}
+
+        Parameters
+        ----------
+        all_metadata: list of xarray.Datasets
+            - The incoming metadata from X number of files
+
+        Returns
+        -------
+        md: dict of xarray Datasets
+            - All metadata merged into a dictionary of xarray Datasets
+        """
+        md = {"METADATA": Dataset()}
+
+        for md_idx in range(len(all_metadata)):
+            # Set required attributes of the top-level metadata when this loop is
+            # started
+            if md_idx == 0:
+                md["METADATA"].attrs = all_metadata[md_idx].attrs
+                md["METADATA"].attrs["source_file_names"] = []
+                md["METADATA"].attrs["source_file_attributes"] = {}
+                md["METADATA"].attrs["source_file_datetimes"] = []
+                md["METADATA"].attrs["end_datetime"] = all_metadata[-1].end_datetime
+
+            # Add to optional attributes of the top-level metadata for each xobj
+            # provided
+            fname = all_metadata[md_idx].attrs["source_file_names"][0]
+            md["METADATA"].attrs["source_file_names"].append(basename(fname))
```

Comment:
Here is where source file names are set for incoming files.

Reply:
I'm wondering if we want source file start datetime, end datetime, and maybe filename date time? Since these are really for reference only, and we don't have a fully documented standard yet I think we do not change anything here, and it should be an easy update once we finalize the standards

Reply:
Talked it over with @cpcamacho , I think this is good for now, but I expect we are going to want to update the xarray_standards to include:

* source_file_start_datetimes (this should correspond to the start_datetime that is in the individual metadata for that file)
* source_file_end_datetimes (this should correspond to the end_datetime that is in the individual metadata for that file)
* source_file_filename_datetimes (this is SPECIFICALLY the datetime in the filename itself, which may or may not match the start datetime.  This is optional.  Sometimes files are WRONG with the datetime that is stored in the filename, so this is really just a sanityh check, that is not really always required)
* source_file_attributes (this is good as is in here)
* source_file_names (this is also good as is)

Once we update the xarray_standards we can revisit this.  Since we don't have the xarray standards 100% finalized, we'll just leave this for now.

Reply:
Sounds good to me!

Reply:
See #817